src/engine/Engine.h

```diff
@@ -367,6 +367,91 @@ class Engine {
     }
   }
 
+  template<typename A, typename B, typename E, int K>
```

Comment:
These are subtree types? Could you give them more descriptive names? Also goes for variable names, e,g, `jcls`